### ISSUE TITLE
Linear Algebra API - Part I : Vector and Matrix

### ISSUE THREAD

**Issue by ratnania** [pyccel/IGA-Python#5]
Up to now, we were relying on `sympy.Matrix` and its variants to handle our Linear Algebra operations.
It is now time to have our own Linear Algebra building blocks, and it starts with the `Vector` and `Matrix` objects.
This issue aims to:

- [ ] have a new API for `Vector` and `Matrix` 
- [ ] remove the use of `Tuple` from all our examples (sympde and psydac)
- [ ] remove the use of `sympy.Matrix`-like objects appearing inside sympde, when possible

The new objects should be used in two different ways:
- [ ] As containers. example:
```python
b = Vector([1.,2.,3.], name='b')
M = Matrix([[1.,2.],[3.,4.]], name='M')
```
- [ ]  as variables. example:
```python
b = Vector(name='b', is_real=True)
M = Matrix(name='M', is_complex=True)
```
in the which case, these variables will appear as free-variables in our expressions (linear, bilinear forms, etc)

**Comment by yguclu** [comment]
I feel that we need to clarify what kind of entries are allowed in a `Vector` or `Matrix`.

If any expression of the spatial coordinates is allowed, then it behaves just like a `Function`, but without specifying the arguments. In fact, it would seem natural to subclass `Function` to get the classes `ScalarFunction`, `VectorFunction`, and `MatrixFunction`.

**Comment by ratnania** [comment]
Yes, we should differentiate between a `(Scalar/)Vector/Matrix` valued functions and other expressions.
This means that we need to check, at the construction, that `Vector/Matrix` expressions do not contain the coordinates.
For the `(Scalar/)Vector/Matrix` valued functions, we'll move the discussion to the associated issue.
